runner/src/recipe-manager.ts

```diff
@@ -0,0 +1,327 @@
+/**
+ * RecipeManager: Unified Recipe Storage and Sync System
+ *
+ * Design goals:
+ * 1. Single storage model: Uses cells in the storage layer for persistent storage
+ * 2. Preserves recipe IDs: Maintains consistency between local and remote IDs
+ * 3. Clear publishing flow: Only syncs with Blobby when explicitly requested
+ * 4. Attempts to download recipes from Blobby if no cell is found
+ * 5. Minimize requirements that Blobby is available for a space to run recipes
+ *
+ * Storage layers:
+ * - In-memory cache: Fast access during runtime
+ * - Cell storage: Persistent local storage
+ * - Blobby storage: Remote storage for sharing recipes
+ *
+ * Usage:
+ * - Use the singleton instance exported as `recipeManager`
+ * - For new code, prefer the `recipeManager` object
+ */
+
+import { JSONSchema, Module, Recipe, Schema } from "@commontools/builder";
+import { storage } from "./storage.ts";
+import { Cell } from "./cell.ts";
+import { createRef } from "./doc-map.ts";
+import { getCell } from "./cell.ts";
+import { runtime } from "@commontools/runner";
+import { getBlobbyServerUrl } from "./blobby-storage.ts";
+
+const inProgressCompilations = new Map<string, Promise<Recipe>>();
+
+// Schema definitions
+export const recipeMetaSchema = {
+  type: "object",
+  properties: {
+    id: { type: "string" },
+    src: { type: "string" },
+    spec: { type: "string" },
+    parents: { type: "array", items: { type: "string" } },
+    recipeName: { type: "string" },
+  },
+  required: ["id"],
+} as const satisfies JSONSchema;
+
+export type RecipeMeta = Schema<typeof recipeMetaSchema>;
+
+// Type guard to check if an object is a Recipe
+function isRecipe(obj: Recipe | Module): obj is Recipe {
+  return "result" in obj && "nodes" in obj;
+}
+
+// FIXME(ja): what happens when we have multiple active spaces... how do we make
+// sure we register the same recipeMeta in multiple spaces?
+const recipeMetaMap = new WeakMap<Recipe, Cell<RecipeMeta>>();
+const recipeIdMap = new Map<string, Recipe>();
```

Comment:
Might be better to wrap these as members of the class rather than in the module? It makes multiple-instances of `RecipeManager` share the same cache which is surprising behaviour, to me.

Reply:
my original thought was the same as you.  I switched when I realized a recipe is unique (hopefully) by ID.  And thus recompiling it shouldn't change it.  

the "MetaMap" is going away once we make the RecipeMeta a LinkedCell in the Charm Schema.  Then all the "load this / save this / ..." go away.. loading the charm gives you the these attributes

Reply:
so I think we can keep them as is?